### PR TITLE
Implement staking mechanism

### LINKED ISSUE #68: write detailed imeplementation for staking of tokens to become a validator
Create a function that allows users to stake tokens and become validators. 

For now, the minimum stake required to become a validator is 200 Predifi tokens. 

A user must stake 200 Predifi tokens to be eligible for the validator role. 

Use OpenZeppelin's access control to assign the validator role to the user. 

write detailed tests covering all implementations, including tests for edge cases associated with the functionality.

eta 2 days

### PR DESCRIPTION
**Pull request by codebestia**
Closes #68 

## Summary
This PR implements functionality for users to stake tokens and become validators in the system. Users who stake at least 200 Predifi tokens are granted the validator role using OpenZeppelin's access control.

## Features
- Added `stake` function that:
  - Allows users to stake Predifi tokens against a specific pool
  - Enforces minimum stake requirement (200 tokens)
  - Automatically grants validator role upon meeting requirements
  - Supports adding to existing stakes
  - Emits an event when staking occurs

## Technical Details
- Implemented minimum stake validation with clear error message
- Uses OpenZeppelin's access control to manage validator roles
- Stores stake amounts in a mapping of (pool_id, user_address) → amount
- Tracks cumulative stakes when users add to existing positions

## Testing
- Test case covers function behavior with min stake amount.
- Staking below minimum amount raises error.
- Validator role assignment verification

## Additional Context
- Fixed pool count test by incrementing pool count by 1 after  successful creation of a pool.

### CODE REVIEW

**Comment by codebestia** [comment]
@Akshola00 Please review

**Comment by Yunusabdul38** [comment]
Nice work @codebestia 
resolve the conflict in your branch

File: contract/.tool-versions [hunk 1]
```diff
@@ -1,2 +1,3 @@
-scarb 2.11.3
-starknet-foundry 0.39.0
\ No newline at end of file
+scarb 2.9.2
```
Comment (Akshola00) [inline contract/.tool-versions:1]: leave theversion as it is 
scarb 2.11.3 and starknet-foundry 0.39.0

File: contract/src/predifi.cairo [hunk 2]
```diff
@@ -21,14 +23,38 @@ pub mod Predifi {
...
         pools: Map<u256, PoolDetails>, // pool id to pool details struct
         pool_count: u256, // number of pools available totally
         pool_odds: Map<u256, PoolOdds>,
         pool_stakes: Map<u256, UserStake>,
         pool_vote: Map<u256, bool>, // pool id to vote
         user_stakes: Map<(u256, ContractAddress), UserStake>, // Mapping user -> stake details
+        #[substorage(v0)]
+        pub accesscontrol: AccessControlComponent::Storage,
+        #[substorage(v0)]
+        src5: SRC5Component::Storage,
```
Comment (Akshola00) [inline contract/src/predifi.cairo:57]: create an array for all validators
for everyone that is a validator, the contract address should be added to the array

**Review by Akshola00 — CHANGES_REQUESTED** [review]

**Comment by codebestia** [comment]
@Akshola00 @Yunusabdul38 
I have fixed the conflicts and implemented the changes.
Please review.

**Comment by codebestia** [comment]
The failed test is not from my changes.

**Comment by Akshola00** [comment]
@codebestia pls ix failing test

run 
```
scarb fmt --check
scarb fmt
scarb build
snforge test
```

and confirm no errors

**Comment by Akshola00** [comment]
> The failed test is not from my changes.

Thanks for clarifying will look into it

**Comment by Akshola00** [comment]
Lgtm

**Comment by codebestia** [comment]
> > The failed test is not from my changes.
> 
> Thanks for clarifying will look into it

I fixed it already. it was a minor issue.
